### ISSUE TITLE
Can't use regex modifiers with the pattern function in YAML rulesets

### ISSUE THREAD

**Issue by arno-di-loreto** [stoplightio/spectral#242]
### **I'm submitting a...**
  - [x] bug report
  - [ ] feature request

### What is the current behavior?

When using the pattern function in a YAML ruleset with spectral-cli, regex with modifiers (`/^.*dto$/i` for example) do not work.
They work when using the spectral module and defining ruleset in node.

YAML ruleset and swagger 2.0 YAML file to reproduce the bug can be found here: https://gist.github.com/arno-di-loreto/63c307e14d8a01928b98625191d59bb7

Not working in YAML ruleset
```
rules:
  schema-name-is-valid-not-working:
    summary: Definition name must not use dto suffix
    given: $.definitions
    severity: 0
    then:
      field: "@key"
      function: pattern
      functionOptions:
        notMatch: /^.*dto$/i
```

Working in node:
```
const rules = () => {
  return { 
    'definition-name-is-valid': {
      summary: 'Definition name must not use dto suffix',
      given: '$.definitions',
      type: RuleType.STYLE,
      severity: DiagnosticSeverity.Error,
      then: {
        field: '@key',
        function: RuleFunction.PATTERN,
        functionOptions: {
           notMatch: /^.*dto$/i
        }
      }
    }
}
```

### What is the expected behavior?

Being able to use regex with modifiers with the pattern function (for both match and notMatch options) in YAML rulesets.

### Please tell us about your environment:

  - spectral-cli -v: @stoplight/spectral/3.0.3 win32-x64 node-v8.11.3

### Other information

I think that when the ruleset is loaded, the /^.*dto$/i regex is loaded as a simple string and not a regex.

Proposed modification in  https://github.com/stoplightio/spectral/blob/master/src/functions/pattern.ts

- Analyzing the match and notMatch regexes to extract regex and modifiers
- Then use new Regex(regex, modifiers)

(Idea based on this stackoverflow response: https://stackoverflow.com/a/874742, I'm not sure about the regex to parse regex, maybe a simple substring would be simpler)

I think I can do it myself and propose a pull request if it is actually a bug (I may have miss something)

**Comment by P0lip** [comment]
@arno-di-loreto 

It's a bug - we should be supporting all possible modifiers, as they are very commonly used. Unfortunately, JS regular expressions cannot be represented in any way in YAML or JSON format, we need to construct a regular expression during the runtime execution based on provided input. 
The proposed modification sounds reasonable to me, so don't hesitate, go ahead and tackle the issue if you'd like to!
Another idea we could consider is using a YAML sequence (arrays in JSON) and having something looking more or less as follows:
```ts
notMatch: [^.*dto$, i]
```
then, in pattern.ts, we could have a very simple condition, such as
```ts
const re = Array.isArray(notMatch) ? new RegExp(...notMatch) : new RegExp(match);
```
Obviously, `IRulePatternOptions` has to be updated as well
```ts
export interface IRulePatternOptions {
  /** regex that target must match */
  match?: RegExp | string | [string, void | string];

  /** regex that target must not match */
  notMatch?: RegExp | string | [string, void | string];
}
```

The above would save us from replacing slashes yet comes with a notable downside as the syntax differs from what we've got in JS.

**Comment by arno-di-loreto** [comment]
For the sake of consistency and simplicity from the users point of view, I would prefer to propose the possibility of having notMatch: /^.*dto$/i
I'll work on it this week.

**Comment by P0lip** [comment]
> For the sake of consistency and simplicity from the users point of view, I would prefer to propose the possibility of having notMatch: /^.*dto$/i

That makes sense to me.
Thank you for taking this issue.
Let me know if there is anything I could help you with.

**Comment by arno-di-loreto** [comment]
Fixed in 3.0.5 🎉
